lib/runners/processor/fxcop.rb

```diff
@@ -20,68 +16,46 @@ def analyzer_version
       ENV.fetch('FXCOP_VERSION')
     end
 
-    def setup
-      # restore project (install dependencies)
-      _, _, status = capture3('dotnet', 'restore')
-      unless status.success?
-        msg = <<~EOS
-          Failed to restore .NET Core Project. You have to put project file (.csproj) under the analysis root. Please check the project structure and 'root_dir' parameter in #{config.path_name}.
-          See: https://help.sider.review/getting-started/custom-configuration#root_dir-option
-          Note: We only support a project managed with MSBuild. No support for other build systems. (e.g. Nuke, Cake)
-        EOS
-        return Results::Failure.new(guid: guid, message: msg, analyzer: analyzer)
-      end
-
-      # install FxCop Analyzers from nuget Repository
-      capture3!('dotnet', 'add', 'package', 'Microsoft.CodeAnalysis.FxCopAnalyzers', '--version', analyzer_version)
-
-      yield
-    end
-
     def analyze(changes)
-      # run 'dotnet build' with static analysis module
       output_file = Tempfile.create(["fxcop-", ""]).path
-      capture3!('dotnet', 'build', '--no-incremental', "-property:errorlog=#{output_file}")
-      output_json = read_output_json(output_file)
+      _, err, status = capture3('Sider.RoslynAnalyzersRunner',
+        '--outputfile', output_file,
+        *changes
+          .changed_paths
+          .select{|p| p.extname.eql?(".cs")}
+          .map{|p| relative_path(working_dir / p, from: current_dir)}
```

Comment:
Currently,  I cannot run smoke tests on my machine, but I think this `working_dir` seems unnecessary. What do you think?

```suggestion
          .map{ |path| relative_path(path, from: current_dir) }
```

Reply:
Let me take an example.
source file: /a/b/c/project/src/hello.cs
the project file directory which contains sider.yml: /a/b/c/project

If the `root_dir` is set to `src` in the sider.yml located in `/a/b/c/project/`, the `current_dir` is set to `/a/b/c/project/src`.
On the other hand, the `changes.changed_paths` file is still `src/hello.cs` which is relative to the `working_dir` (e.g. /a/b/c/project).
I need to pass "hello.cs" to Sider.RoslynAnalyzersRunner because it runs on the `current_dir` (/a/b/c/project/src).
So I normalized the path as follows:
relative_path(working_dir / p, from: current_dir)
->
relative_path("/a/b/c/project" / "src/hello.cs", from: "/a/b/c/project/src") leads to "hello.cs".

Reply:
Yes, I know the parameter of `from` is necessary, but passing the path prefixed `working_dir` seems unnecessary because `p` is already an instance of `Pathname`.

Reply:
Although `p` is an instance of the Pathname class, it has a relative path to the `working_dir`. I just wanted to obtain a relative path to the `current_dir`, so I needed to expand `p` into the absolute path by adding the prefix `working_dir` before splitting into `current_dir` and its relative path. For example, `relative_path("/a/b/c/project" / "src/bar/hello.cs", from: "/a/b/c/project/src")` leads to "bar/hello.cs".

```
changes.rb
def self.calculate(base_dir:, head_dir:, working_dir:, patches:)
  relative_path = working_path.relative_path_from(working_dir)
  changed_paths << relative_path

processor.rb
def relative_path(original, from: working_dir)
  path = Pathname(original)
  if path.relative?
    path = (current_dir / path).cleanpath
  path.relative_path_from(from)
```

Reply:
:thinking: Whichever the `p` is a relative path or not, using `from` of `Processor#relative_path` seems to resolve your demand.

I created a test `test_relative_path` and get the succeeded result.... Do I misunderstand?

```diff
diff --git a/test/processor_test.rb b/test/processor_test.rb
index 3d3ba11..5cf7f70 100644
--- a/test/processor_test.rb
+++ b/test/processor_test.rb
@@ -31,6 +31,20 @@ class ProcessorTest < Minitest::Test
     )
   end

+  def test_relative_path
+    with_workspace do |workspace|
+      working_dir = workspace.working_dir
+      (working_dir / "src").mkdir
+      (working_dir / "src/abc.cs").write("abc")
+      (working_dir / "src/foo.cs").write("foo")
+
+      current_dir = (working_dir / "src").realpath
+      processor = new_processor(workspace: workspace)
+      assert_equal "src/abc.cs", processor.relative_path(Pathname("src/abc.cs"), from: working_dir).to_path # <-- This line is the case `root_dir` is set as `"src"`
+      assert_equal "abc.cs", processor.relative_path(Pathname("src/abc.cs"), from: current_dir).to_path
+    end
+  end
+
   def test_capture3_env_setup
     with_workspace do |workspace|
       mock(Open3).capture3({"RUBYOPT" => nil, "GIT_SSH" => (workspace.working_dir / "id_rsa").to_s},
```

Reply:
:thinking: Whichever the `p` is a relative path or not, using `from` of `Processor#relative_path` seems to resolve your demand.

I created a test `test_relative_path` and get the succeeded result.... Do I misunderstand?

```diff
diff --git a/test/processor_test.rb b/test/processor_test.rb
index 3d3ba11..5cf7f70 100644
--- a/test/processor_test.rb
+++ b/test/processor_test.rb
@@ -31,6 +31,20 @@ class ProcessorTest < Minitest::Test
     )
   end

+  def test_relative_path
+    with_workspace do |workspace|
+      working_dir = workspace.working_dir
+      (working_dir / "src").mkdir
+      (working_dir / "src/abc.cs").write("abc")
+      (working_dir / "src/foo.cs").write("foo")
+
+      current_dir = (working_dir / "src").realpath
+      processor = new_processor(workspace: workspace)
+      assert_equal "src/abc.cs", processor.relative_path(Pathname("src/abc.cs"), from: working_dir).to_path # <-- This line is the case `root_dir` is set as `"src"`
+      assert_equal "abc.cs", processor.relative_path(Pathname("src/abc.cs"), from: current_dir).to_path
+    end
+  end
+
   def test_capture3_env_setup
     with_workspace do |workspace|
       mock(Open3).capture3({"RUBYOPT" => nil, "GIT_SSH" => (workspace.working_dir / "id_rsa").to_s},
```

Reply:
To update the `current_dir` value in the `processor` instance, you need to call the `push_dir` method. Try the code below, you will get assertion errors.
```
      processor = new_processor(workspace: workspace)
      processor.push_dir(current_dir) do
        assert_equal "src/abc.cs", ...
        assert_equal "abc.cs", ...
      end
```

Reply:
OK. I understood. You're right. Sorry to bother you.